File: py/Permutations.py
class Solution:
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        Solution.perm_recur(nums, 0, res)
        return res

    @staticmethod
    def perm_recur(nums, start, res):
        if start == len(nums) - 1:
            res.append(nums[:])
        else:
            for i in range(start, len(nums)):
                nums[start], nums[i] = nums[i], nums[start]
                Solution.perm_recur(nums, start + 1, res)
                nums[start], nums[i] = nums[i], nums[start]

File: cpp/Permutations.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        backtrack(0, nums, res);
        return res;
    }
private:
    void backtrack(int idx, vector<int>& nums, vector<vector<int>> &res) {
        if (idx == nums.size()) {
            res.push_back(nums);
            return;
        }
        for (int i = idx; i < nums.size(); i ++) {
            swap(nums[i], nums[idx]);
            backtrack(idx + 1, nums, res);
            swap(nums[i], nums[idx]);
        }
    }
};

int main() {
    Solution s;
    vector<int> nums = {1,2,3};
    auto res = s.permute(nums);
    for (auto &v: res) {
        cout << '[';
        for (int n: v) {
            cout << n << ' ';
        }
        cout << ']' << endl;
    }
}
